python/tank/descriptor/io_descriptor/shotgun_entity.py

```diff
@@ -208,3 +232,76 @@ def get_latest_version(self, constraint_pattern=None):
         log.debug("Latest version resolved to %s" % desc)
         return desc
 
+    def get_latest_cached_version(self, constraint_pattern=None):
+        """
+        Returns a descriptor object that represents the latest version
+        that is locally avaiable in the bundle cache search path.
+
+        :param constraint_pattern: If this is specified, the query will be constrained
+               by the given pattern. Version patterns are on the following forms:
+
+                - v0.1.2, v0.12.3.2, v0.1.3beta - a specific version
+                - v0.12.x - get the highest v0.12 version
+                - v1.x.x - get the highest v1 version
+
+        :returns: instance deriving from IODescriptorBase or None if not found
+        """
+        if constraint_pattern:
+            log.warning("%s does not support version constraint patterns." % self)
+
+        log.debug("Looking for cached versions of %r..." % self)
+        # for each of the cache paths, look one level above
+        # and enumerate all items
+        all_versions = []
+
+        for possible_cache_path in self._get_cache_paths():
+            parent_folder = os.path.dirname(possible_cache_path)
+            log.debug("looking in '%s'" % parent_folder)
+            if os.path.exists(parent_folder):
+                for version_folder in os.listdir(parent_folder):
```

Comment:
same here. Wondering if this repeated logic is similar enough to extract and share?